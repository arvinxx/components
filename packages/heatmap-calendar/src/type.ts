/**
 * 热力图数据结构
 */
export interface ItemData {
  /**
   * 日期
   */
  date: string;
  /**
   * 天数
   */
  day: number;
  /**
   * 月份
   */
  month: number;
  /**
   * 热点数
   */
  total: number;
  /**
   * 周
   */
  week: number;
  /**
   * 年
   */
  year: number;
}

export type HeatmapCalendarData = ItemData[];
