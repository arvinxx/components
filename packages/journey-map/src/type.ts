/**
 * 阶段
 */
export interface Step<T = string> {
  /**
   * 阶段id
   * @TJS-type string
   */
  id: T;
  /**
   * 阶段名称
   */
  name: string;
  /**
   * 该阶段的颜色
   */
  color?: string;
}

/**
 * 用户旅程地图数据
 */
export interface JourneyMapData<T extends string = string> {
  /**
   * 大的步骤
   */
  steps: Step<T>[];
  /**
   * 用户行为
   */
  actions: Record<T, UserAction[]>;
}

/**
 * 用户行为
 */
export interface UserAction {
  /**
   * 行为名称
   */
  title: string;
  /**
   * 说明
   */
  description?: string;

  /**
   * 有关的想法
   */
  thoughts?: string[];
  /**
   * 情绪得分
   */
  emotion: number;
  /**
   * 痛点
   */
  painSpot?: string[];
}

/**
 * 解析后的 YML 数据
 */
export interface JourneyMapYML<T extends string = string> {
  /**
   * 步骤
   */
  steps: Step<T>[];
  /**
   * 用户行为
   */
  actions: Record<T, Record<T, number>[]>;
}
