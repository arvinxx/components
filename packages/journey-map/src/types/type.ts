/**
 * 阶段
 */
export interface Stage<T = string> {
  /**
   * 阶段id
   * @TJS-type string
   */
  id: T;
  /**
   * 阶段名称
   */
  name: string;
  /**
   * 该阶段的颜色
   */
  color?: string;
}

/**
 * 用户旅程地图数据
 */
export interface JourneyMapData<T extends string = string> {
  /**
   * 大的步骤
   */
  stages: Stage<T>[];
  /**
   * 用户行为
   */
  actions: Record<T, UserAction[]>;
}

/**
 * 用户行为类型
 */
export interface UserAction {
  /**
   * 行为名称
   */
  name: string;
  /**
   * 情绪得分
   * @minimum -2
   * @maximum 2
   * @TJS-type integer
   */
  emotion?: number;

  /**
   * 说明
   */
  description?: string;

  /**
   * 想法
   */
  thoughts?: string[];

  /**
   * 痛点
   */
  painSpot?: string[];
}
