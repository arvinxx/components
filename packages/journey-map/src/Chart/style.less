.avx-journey-map-chart {
  display: flex;
}
