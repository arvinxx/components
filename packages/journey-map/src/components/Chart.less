.avx-journey-map-chart {
  display: flex;
  margin-top: 24px;
}
