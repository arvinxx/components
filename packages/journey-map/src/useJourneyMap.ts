import useMergeValue from 'use-merge-value';
import { useEffect, useState } from 'react';

import { getServiceToken, YMLToJSON } from './utils';
import type { YMLJourneyMapData, JourneyMapData } from './types';

interface StoreParams {
  data: JourneyMapData | string;
  onChange?: (data: JourneyMapData) => void;
  title?: string;
}

export const useJourneyMap = ({
  data,
  onChange,
  title: initTitle,
}: StoreParams) => {
  const defaultJourneyMap: JourneyMapData = { stages: [], actions: {} };

  const hook = useMergeValue<JourneyMapData>(
    typeof data === 'string' ? defaultJourneyMap : data,
    { onChange },
  );

  const [title, setTitle] = useState<string>(initTitle);

  const [store, setStore] = hook;
  /**
   * 从 YML URL 中解析 JSON
   * @param url
   */
  const fetchYMLToJSON = async (url: string): Promise<YMLJourneyMapData> => {
    const res = await fetch(url);
    const yml = await res.text();

    return YMLToJSON(yml);
  };

  useEffect(() => {
    if (typeof data === 'string' && data.startsWith('http')) {
      fetchYMLToJSON(data).then((result) => {
        const { stages, actions } = result;
        setTitle(result.title);
        setStore({ stages, actions });
      });
    }
  }, [data]);

  const actionList = Object.values(store.actions).flat();

  return { store, actionList, actionLength: actionList.length, title };
};

export const JourneyMapStore = getServiceToken(useJourneyMap);
