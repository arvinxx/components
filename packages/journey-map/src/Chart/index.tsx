import type { FC } from 'react';
import React, { useContext } from 'react';
import type { AreaConfig } from '@ant-design/charts/es/area';
import { Area } from '@ant-design/charts';
import { blue } from '@ant-design/colors';

import { Section } from '../components';
import { JourneyMapStore } from '../useJourneyMap';
import './style.less';

interface ChartProps {
  color?: string;
}

/**
 * 更
 * @param text
 */
const getFormattedEmotion = (text: string | number) => {
  switch (text.toString()) {
    case '-2':
      return '极差';
    case '-1':
      return '差';
    case '0':
      return '一般';
    case '1':
      return '好';
    case '2':
      return '极好';
    default:
      return text;
  }
};

const Chart: FC<ChartProps> = ({ color = blue[2] }) => {
  const [store] = useContext(JourneyMapStore);
  const { actions } = store;
  const data = Object.values(actions).flat();

  const config: AreaConfig = {
    data,
    padding: [24, 0, 16, 0],
    tooltip: {
      formatter: (datum) => {
        const { emotion } = datum;

        return { name: '情绪值', value: getFormattedEmotion(emotion) };
      },
    },
    xField: 'name',
    yField: 'emotion',
    yAxis: {
      alias: '情绪值',
      label: {
        formatter: getFormattedEmotion,
      },
      grid: {
        line: null,
      },
    },
    xAxis: {
      label: null,
      line: null,
    },

    annotations: [
      {
        type: 'line',
        start: ['min', 'median'],
        end: ['max', 'median'],
        style: {
          stroke: blue[2],
          lineDash: [6, 4],
        },
      },
    ],
    autoFit: true,
    smooth: true,
    areaStyle: () => ({ fill: `l(270) 0:${blue[0]} 1:${blue[4]}` }),

    line: {
      color,
    },

    height: 300,
    animation: { appear: { animation: 'fade-in' } },
  };
  return (
    <div className="avx-journey-map-chart">
      <Section height={300}>体验情绪</Section>
      <div style={{ width: '100%' }}>
        <Area {...config} />
      </div>
    </div>
  );
};

export default Chart;
