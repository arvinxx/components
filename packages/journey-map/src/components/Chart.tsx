import type { FC } from 'react';
import React, { useContext } from 'react';
import type { AreaConfig } from '@ant-design/charts/es/area';
import { Area } from '@ant-design/charts';
import { blue } from '@ant-design/colors';

import Stage from './Stage';
import { JourneyMapStore } from '../useJourneyMap';
import './Chart.less';

interface ChartProps {
  color?: string;
}

/**
 * 更
 * @param text
 */
const getFormattedEmotion = (text: string | number) => {
  switch (text.toString()) {
    case '-2':
      return '极差';
    case '-1':
      return '差';
    case '0':
      return '一般';
    case '1':
      return '好';
    case '2':
      return '极好';
    default:
      return text;
  }
};

const Chart: FC<ChartProps> = ({ color = blue[2] }) => {
  const [store] = useContext(JourneyMapStore);
  const { actions } = store;
  const data = Object.values(actions).flat();

  const config: AreaConfig = {
    data,
    // padding: [24, 40],
    tooltip: {
      formatter: (datum) => {
        const { emotion } = datum;

        return { name: '情绪值', value: getFormattedEmotion(emotion) };
      },
    },
    xField: 'title',
    yField: 'emotion',
    yAxis: {
      alias: '情绪值',
      label: {
        formatter: getFormattedEmotion,
      },
      grid: {
        line: null,
      },
    },
    xAxis: {
      label: null,
      line: null,
    },

    annotations: [
      {
        type: 'line',
        start: ['min', 'median'],
        end: ['max', 'median'],
        style: {
          stroke: blue[2],
          lineDash: [6, 4],
        },
      },
    ],
    autoFit: true,
    smooth: true,
    areaStyle: () => ({ fill: `l(270) 0:${blue[0]} 1:${blue[4]}` }),

    line: {
      color,
    },

    height: 300,
    animation: { appear: { animation: 'fade-in' } },
  };
  return (
    <div className="avx-journey-map-chart">
      <div style={{ width: 40 }}>
        <Stage height={300}>体验情绪</Stage>
      </div>
      <div style={{ marginTop: 2, width: '100%' }}>
        <Area {...config} />
      </div>
    </div>
  );
};

export default Chart;
