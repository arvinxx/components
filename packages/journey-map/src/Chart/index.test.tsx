import React from 'react';
import { render } from '@testing-library/react';

import Chart from './index';
import { JourneyMapStore, useJourneyMap } from '../useJourneyMap';

import type { FC } from 'react';
import type { JourneyMapData } from '../types';

const Demo: FC<{ data: JourneyMapData }> = ({ data }) => {
  const value = useJourneyMap({
    data,
  });
  return (
    <JourneyMapStore.Provider value={value}>
      <Chart />
    </JourneyMapStore.Provider>
  );
};

describe('Chart', () => {
  it('默认状态', async () => {
    const { container } = render(
      <Demo
        data={{
          actions: {
            '1': [{ name: '1', emotion: 1 }],
            '2': [{ name: '2', emotion: 2 }],
            '0': [{ name: '0', emotion: -1 }],
            '3': [{ name: '3', emotion: -2 }],
            '4': [{ name: '4', emotion: 0 }],
            '5': [{ name: '5' }],
          },
          stages: [],
        }}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
