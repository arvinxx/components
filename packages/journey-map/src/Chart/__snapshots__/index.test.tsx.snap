// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart 默认状态 1`] = `
<div>
  <div
    class="avx-journey-map-chart"
  >
    <div
      class="avx-journey-map-section"
      style="height: 300px;"
    >
      <div
        class="avx-journey-map-section-text"
      >
        体验情绪
      </div>
    </div>
    <div
      style="width: 100%; height: 300px;"
    >
      <div
        data-chart-source-type="G2Plot"
        size-sensor-id="1"
        style="height: inherit;"
      >
        <div
          style="position:relative;"
        >
          <canvas
            height="400"
            style="width: 400px; height: 400px; display: inline-block; vertical-align: middle;"
            width="400"
          />
        </div>
        <object
          class="size-sensor-object"
          data="about:blank"
          style="display: block; position: absolute; top: 0px; left: 0px; height: 100%; width: 100%; overflow: hidden; pointer-events: none; z-index: -1; opacity: 0;"
          tabindex="-1"
          type="text/html"
        />
      </div>
    </div>
  </div>
</div>
`;
